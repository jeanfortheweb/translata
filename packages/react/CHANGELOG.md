# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2019-10-17)


### New

* React Integration ([28488ec8326562b0ad5fc26be364e83f7a312b70](https://github.com/jeanfortheweb/translata/commit/28488ec8326562b0ad5fc26be364e83f7a312b70))
