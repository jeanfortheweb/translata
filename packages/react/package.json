{
  "name": "@translata/react",
  "version": "0.0.0",
  "description": "React integration for translata: The Composable Translation Utility",
  "license": "MIT",
  "repository": "github:jeanfortheweb/translata",
  "keywords": [
    "translation",
    "i18n",
    "typescript",
    "functional",
    "react",
    "hooks"
  ],
  "engines": {
    "node": ">= 10.15.3"
  },
  "publishConfig": {
    "access": "public"
  },
  "main": "lib/cjs/index.js",
  "module": "lib/esm/index.js",
  "files": [
    "lib"
  ],
  "scripts": {
    "compile:cjs": "rm -rf lib/cjs && tsc --p tsconfig.cjs.json",
    "compile:esm": "rm -rf lib/esm && tsc --p tsconfig.esm.json"
  },
  "peerDependencies": {
    "react": "^16.10.2"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^4.1.2",
    "@testing-library/react": "^9.3.0",
    "@translata/core": "^0.3.1",
    "@types/react": "^16.9.9",
    "react": "^16.10.2",
    "react-dom": "^16.10.2"
  }
}
