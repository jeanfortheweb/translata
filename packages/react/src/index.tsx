import { Translator } from '@translata/core';
import React, {
  createContext,
  useContext,
  useState,
  useMemo,
  ReactElement,
} from 'react';

/**
 * Create a translator provider, including hooks to consume the translation context.
 *
 * @param defaultLocale Default locale.
 * @param factory Translator factory.
 */
export function createTranslatorProvider<
  TranslatorType extends Translator<any>
>(defaultLocale: string, factory: (locale: string) => TranslatorType) {
  const Context = createContext({
    translator: (null as any) as TranslatorType,
    locale: (null as any) as string,
    setLocale: (null as any) as (locale: string) => void,
  });

  /**
   * The Translator Provider.
   *
   * @param props
   */
  function Provider(props: { children: ReactElement }) {
    const [locale, setLocale] = useState(defaultLocale);
    const translator = useMemo(() => factory(locale), [locale]);
    const state = useMemo(
      () => ({
        translator,
        locale,
        setLocale,
      }),
      [translator, locale],
    );

    return <Context.Provider value={state}>{props.children}</Context.Provider>;
  }

  /**
   * Use the current translator from context.
   */
  function useTranslator(): TranslatorType {
    return useContext(Context).translator;
  }

  /**
   * Use the current locale from context.
   */
  function useLocale(): string {
    return useContext(Context).locale;
  }

  /**
   * Use the locale setter from context, this allows you to change
   * the default locale on context level.
   */
  function useSetLocale(): (value: string) => void {
    return useContext(Context).setLocale;
  }

  return Object.assign(Provider, { useLocale, useSetLocale, useTranslator });
}
