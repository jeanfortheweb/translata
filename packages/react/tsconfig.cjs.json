{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "lib/cjs",
    "module": "commonjs",
    "moduleResolution": "node",
    "declaration": true,
    "jsx": "react"
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["lib/**", "**/__tests__/**"]
}
