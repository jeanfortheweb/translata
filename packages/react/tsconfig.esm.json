{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "lib/esm",
    "target": "es5",
    "module": "esnext",
    "moduleResolution": "node",
    "declaration": true,
    "jsx": "react"
  },
  "include": ["src/**/*.ts", "src/**/*.tsx"],
  "exclude": ["lib/**", "**/__tests__/**"]
}
