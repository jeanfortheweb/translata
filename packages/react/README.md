# @translata/react <!-- omit in toc -->

Provides a react integration for **Translata**.

## Installation <!-- omit in toc -->

```sh
yarn add @translata/core @translata/react
```

## Documentation <!-- omit in toc -->

- [createTranslatorProvider](#createtranslatorprovider)

### createTranslatorProvider

Create a translator provider using a default locale and a factory function.
The factory function gets the current active locale and should return a setup translator function which is created by [createTranslator](../core/README.md#createtranslator):

```ts
// AppTranslator.ts
import { createTranslator, withTranslations, withDefaultLocale } from '@translata/core';
import { createTranslatorProvider } from '@translata/react';

export default createTranslatorProvider('en', locale => createTranslator(
  withTranslations('en', {
    'message.greeting': 'Hello!',
    'button.de': 'Goto german',
    'button.en': 'Goto english'
  }),
  withTranslations('en', {
    'message.greeting': 'Hallo!',
    'button.de': 'Wechsel zu Deutsch',
    'button.en': 'Wechsel zu Englisch'
  })
  withDefaultLocale(locale)
));
```

Wrap this Provider around your application:

```tsx
// App.tsx
import AppTranslator from './AppTranslator';
import Component from './Component';

export default () => {
  return (
    <AppTranslator>
      <Component />
    </AppTranslator>
  );
}
```

Use the translator context inside your components:

```tsx
// Component.tsx
import AppTranslator from './AppTranslator';

export default () => {
  // get the actual translator function.
  const _ = AppTranslator.useTranslator();

  // get the current active locale.
  const locale = AppTranslator.useLocale();

  // get a locale setter to change the locale on the context.
  // this will reinvoke your translator factory function passed to createTranslatorProvider().
  const setLocale = AppTranslator.useSetLocale();

  // callback to change locale to "de"
  const setLocaleToDe = useCallback(() => {
    setLocale('de');
  }, [])

  // callback to change locale to "en"
  const setLocaleToEn = useCallback(() => {
    setLocale('en');
  }, [])

  // use the translata as you would without react.
  return (
    <div>
      {_('message.greeting')}
      <button disabled={locale === 'de'} onClick={setLocaleToDe}>{_('button.de')}</button>
      <button disabled={locale === 'en'} onClick={setLocaleToDe}>{_('button.en')}</button>
    </div>
  );
}
```
