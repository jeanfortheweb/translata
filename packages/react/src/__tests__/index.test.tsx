import React, { ReactElement, useEffect } from 'react';
import { createTranslatorProvider } from '..';
import { render, act, getByTestId } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {
  createTranslator,
  withTranslations,
  withDefaultLocale,
} from '@translata/core';

const Translator = createTranslatorProvider('en', locale =>
  createTranslator(
    withTranslations('en', {
      greeting: 'Hello User',
    }),
    withTranslations('de', {
      greeting: 'Hallo Benutzer',
    }),
    withDefaultLocale(locale),
  ),
);

const Root = ({ children }: { children: ReactElement }) => {
  return <Translator>{children}</Translator>;
};

it('should setup a working provider', async () => {
  const Component = () => {
    const _ = Translator.useTranslator();

    return <div data-testid="div">{_('greeting')}</div>;
  };

  act(() => {
    render(
      <Root>
        <Component />
      </Root>,
    );
  });

  const element = getByTestId(document.body, 'div');

  expect(element).toMatchSnapshot('english');
});

it('should change the locale', async () => {
  const Component = () => {
    const _ = Translator.useTranslator();
    const setLocale = Translator.useSetLocale();

    useEffect(() => setLocale('de'), []);

    return <div data-testid="div">{_('greeting')}</div>;
  };

  act(() => {
    render(
      <Root>
        <Component />
      </Root>,
    );
  });

  const element = getByTestId(document.body, 'div');

  expect(element).toMatchSnapshot('german');
});

it('should change provide the current locale', async () => {
  const Component = () => {
    const locale = Translator.useLocale();

    return <div data-testid="div">{locale}</div>;
  };

  act(() => {
    render(
      <Root>
        <Component />
      </Root>,
    );
  });

  const element = getByTestId(document.body, 'div');

  expect(element).toMatchSnapshot('locale');
});
