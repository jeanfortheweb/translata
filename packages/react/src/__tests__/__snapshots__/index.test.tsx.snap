// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change provide the current locale: locale 1`] = `
<div
  data-testid="div"
>
  en
</div>
`;

exports[`should change the locale: german 1`] = `
<div
  data-testid="div"
>
  Hallo Benutzer
</div>
`;

exports[`should setup a working provider: english 1`] = `
<div
  data-testid="div"
>
  Hello User
</div>
`;
